gdal/alg/gdalwarpkernel.cpp

```diff
@@ -5821,49 +5831,38 @@ static void GWKAverageOrModeThread( void* pData)
         }
         else
         {
-            nAlgo = GWKAOM_Fmode;
+            nAlgo = AOM::FloatMode;
 
-            if( nSrcXSize > 0 && nSrcYSize > 0 )
-            {
-                pafRealVals = static_cast<float *>(
-                    VSI_MALLOC3_VERBOSE(nSrcXSize, nSrcYSize, sizeof(float)));
-                panRealSums = static_cast<int *>(
-                    VSI_MALLOC3_VERBOSE(nSrcXSize, nSrcYSize, sizeof(int)));
-                if( pafRealVals == nullptr || panRealSums == nullptr )
-                {
-                    VSIFree(pafRealVals);
-                    VSIFree(panRealSums);
-                    return;
-                }
-            }
+            realVals.resize(nSrcXSize * nSrcYSize);
```

Comment:
nSrcXSize * nSrcYSize could in extreme cases overflow int. So we'd need to cast to size_t before multiplying ( I don't think that the multiplication can overflow on 32 bit since we have allocated at least that before, so it would have failed ), and catch a potential exception for the resize

Reply:
@abellgithub ping, just to make sure you saw my above comment

Reply:
I understand the overflow concern.

I don't understand the bit in parens.

Reply:
static_cast<size_t>(nSrcXSize) * nSrcYSize  , on a 32 bit build, could theoretically overflow size_t = uint32_t . But given that it has already been checked previously in https://github.com/OSGeo/gdal/blob/master/gdal/alg/gdalwarpoperation.cpp#L1833 , no need to do it again

Reply:
Oh. Well, the overflow isn't really the issue at that point. If your raster size is larger than 16 exabytes, I don't think this is the appropriate application for you ;)

Reply:
but on a 64 bit build, static_cast<size_t>(nSrcXSize) * nSrcYSize is needed for folks using big -wm settings

Reply:
Oh, size_t is 32-bin on Win32. Do you support 32-bit windows systems?

Reply:
> Oh, size_t is 32-bin on Win32. Do you support 32-bit windows systems?

yes, as well as any other more relevant 32-bit platforms like arm 32 bit builds, etc..  But the cast to size_t is needed mostly for 64 bit builds, because nSrcXSize and nsrcYSize are (32 bit) int.

Reply:
Yes, but then the cast won't buy you anything because size_t is uint32_t and will overflow.  Need to explicitly use a cast to 64-bit type.

I'm not disagreeing that it needs fixing.

Reply:
> size_t is uint32_t and will overflow

no, size_t is 64 bit on 64 bit platforms (at least all those I'm aware of)

Reply:
I was talking about 32 bit, where size_t is also 32 bit, but then the max size of a vector has the same limit.

Anyway, I'll make the change in a bit.